### PR TITLE
Add rtl functionality to keyboard dir mixin

### PR DESCRIPTION
**Pull request by yuriy-fix**
Connected to https://github.com/vaadin/vaadin-list-mixin/issues/65

### CODE REVIEW

**Comment by web-padawan** [comment]
Discussed in person and agreed on the following modifications:

1. Add `_isVertical` method to `KeyboardDirectionClass` (false by default)

2. Update `_isPrevKey` and `_isNextKey` to only use one direction (not both)

3. Update tests to replace `kdm-element` with two classes:

  - `kdm-vertical-element` that will override `_isVertical` to return `true`
  - `kdm-horizontal-element` that will not override`_isVertical`

This would allow us to avoid having `orientation-mixin` dependency here.

It would also mean that we can override `_isVertical` in Accordion (always `true`)  and Tabs (depends on `orientation` property) and remove overrides for `_isPrevKey` / `_isNextKey`.

File: packages/keyboard-direction-mixin/test/keyboard-direction-mixin.test.ts [hunk 1]
```diff
@@ -25,11 +25,41 @@ class KdmElement extends KeyboardDirectionMixin(SlottedItemsMixin(LitElement)) {
...
+  }
+
+  _isVertical() {
+    return false;
+  }
 }
 
 describe('KeyboardDirectionMixin', () => {
-  let element: KdmElement;
+  let element: KdmElementVertical;
```
Comment (web-padawan) [inline packages/keyboard-direction-mixin/test/keyboard-direction-mixin.test.ts]: Let's create two sub-suites:

- `vertical mode` 
- `horizontal mode`

Inside of those there should be `default` and `RTL` sub-suites.

File: packages/keyboard-direction-mixin/test/keyboard-direction-mixin.test.ts [hunk 2]
```diff
@@ -6,8 +6,8 @@ import { KeyboardDirectionMixin } from '../keyboard-direction-mixin';
 
 const { expect } = chai;
 
-@customElement('kdm-element')
-class KdmElement extends KeyboardDirectionMixin(SlottedItemsMixin(LitElement)) {
+@customElement('kdm-element-vertical')
```
Comment (web-padawan) [inline packages/keyboard-direction-mixin/test/keyboard-direction-mixin.test.ts]: Let's rename to `kdm-vertical-element` to follow the covention used in tests (end with `-element`).

File: packages/keyboard-direction-mixin/test/keyboard-direction-mixin.test.ts (hunk 1 above)
Comment (yuriy-fix) [reply packages/keyboard-direction-mixin/test/keyboard-direction-mixin.test.ts]: Done

File: packages/keyboard-direction-mixin/test/keyboard-direction-mixin.test.ts (hunk 2 above)
Comment (yuriy-fix) [reply packages/keyboard-direction-mixin/test/keyboard-direction-mixin.test.ts]: Done

**Comment by web-padawan** [comment]
One idea which came to my mind: we should change how mixins depend on each other.

I changed `SelectionInViewMixin` to not depend on `OrientationMixin` but in fact it shouldn't depend on `KeyboardDirectionMixin` either (it does because of `_isVertical`).

We should instead have other mixin, say `DirectionMixin` with a single `protected get _vertical`.
That would be an "internal" mixin: it means that mixins would apply it internally.

When using all of these mixins together, every of those would apply this common mixin. In order to prevent polluting the prototype chain, we can use [dedupeMixin](https://github.com/ing-bank/lion/blob/9688007123debe900ba599976d1f48d356609d07/packages/core/src/dedupeMixin.js) helper (also used by mixins from Polymer).

**Comment by web-padawan** [comment]
Updated the PR as described above.

- `OrientationMixin` now depends on `DirectionMixin` and overrides `_vertical`
- `SelectionInVIewMixin` still depends on `KeyboardDirectionMixin` because of `_focus` 

Note: the `_vertical` getter in `OrientationMixin` is actually tested through `SelectionInVIewMixin` test suite, which is still not ideal. I couldn't quickly come up with other way of testing it.

File: packages/direction-mixin/direction-mixin.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import { DirectionClass } from './direction-class';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Constructor<T = object> = new (...args: any[]) => T;
+
+const appliedMixins = new WeakMap();
```
Comment (yuriy-fix) [inline packages/direction-mixin/direction-mixin.ts]: Do we really need a `WeakMap` over here?
If I understood correctly, we use only one `DirectionMixin`  and a list of extended classes? 🤔

Comment (web-padawan) [reply packages/direction-mixin/direction-mixin.ts]: Changed to `Set`, thanks.

**Pull request merged**